cmd/dump.go

```diff
@@ -6,22 +6,152 @@
 package cmd
 
 import (
+	"bufio"
+	"bytes"
+	"encoding/json"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"path"
 	"path/filepath"
+	"strings"
+	"syscall"
 	"time"
 
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
+	"gitea.com/macaron/session"
 
-	"github.com/unknwon/cae/zip"
+	archiver "github.com/mholt/archiver/v3"
 	"github.com/unknwon/com"
 	"github.com/urfave/cli"
 )
 
+type dumpFileInfo struct {
+	name string
+	size int64
+}
+
+func (fi dumpFileInfo) Name() string       { return fi.name }
+func (fi dumpFileInfo) Size() int64        { return fi.size }
+func (fi dumpFileInfo) Mode() os.FileMode  { return 0644 }
+func (fi dumpFileInfo) ModTime() time.Time { return time.Now() }
+func (fi dumpFileInfo) IsDir() bool        { return false }
+func (fi dumpFileInfo) Sys() interface{} {
+	return &syscall.Stat_t{Uid: uint32(os.Getuid()), Gid: uint32(os.Getgid())}
+}
+
+func addFileFromBuffer(w archiver.Writer, filePath string, buffer *bytes.Buffer, verbose bool) error {
```

Comment:
how about adding a `addFileFromReader` to allow us to avoid reading the whole db into memory?

Reply:
Indeed but writing the file to a archive requires to know the size of the file previous to providing the data.
It would be possible to seek back to the header and write the size afterwards, but this won't work when writing to stdout.

Reply:
So I think it might be better to revert this commit and still use a temporary file?!

Reply:
Why does it require the size?

Reply:
I don't know exactly for zip. But for tar the size of the file is part of the file header.
And this size is used on exctraction (or any read) to know where the next file header starts.
So supplying an invalid size for a file will result in a corrupt tar file.

Reply:
sigh... I guess in the case of tar you need to send it to a tempfile - but we should default to using a system temp file as per ioutil.TempFile.

Reply:
After a bit of research it turns out that zip uses a file header which may contain the file size and a central directory at the end of the file which must contain the file size.
This works fine when streaming files as the size might be omitted in the file header, but can be remembered and written to the directory.

But this prevents files written to stdout being extracted on the fly using unzip because it needs the central directory from the end of the stream :/

So I think you are right that it would be the best to just use a ioutil.TempFile.